### PR TITLE
ARROW-1882: [C++] Reintroduce DictionaryBuilder

### PR DESCRIPTION
**Pull request by xhochy**
Readded the previous code and moved some small parts to a new common place to share logic between kernels and builder.

### CODE REVIEW

**Comment by wesm** [comment]
Thanks @xhochy! I will review

**Comment by xhochy** [comment]
Green AppVeyor: https://ci.appveyor.com/project/xhochy/arrow/build/1.0.526

**Comment by xhochy** [comment]
@wesm it did not look at first glance that we could easily share more code. I would like to see this merged before 0.8 and then refactor a bit later on to reduce code duplication.

File: cpp/src/arrow/array-test.cc [hunk 1]
```diff
@@ -1619,6 +1619,353 @@ TEST_F(TestAdaptiveUIntBuilder, TestAppendVector) {
...
+  NumericBuilder<TypeParam> builder;
+  // DictionaryBuilder<TypeParam> builder;
+  ASSERT_OK(builder.Append(static_cast<typename TypeParam::c_type>(1)));
+  ASSERT_OK(builder.Append(static_cast<typename TypeParam::c_type>(2)));
+  ASSERT_OK(builder.Append(static_cast<typename TypeParam::c_type>(1)));
+
+  std::shared_ptr<Array> intermediate_result;
+  ASSERT_OK(builder.Finish(&intermediate_result));
+  DictionaryBuilder<TypeParam> dictionary_builder(default_memory_pool());
+  ASSERT_OK(dictionary_builder.AppendArray(*intermediate_result));
```
Comment (wesm) [inline cpp/src/arrow/array-test.cc]: We might consider removing these `AppendArray` methods in favor of the kernel-based approach

**Review by wesm — APPROVED** [review]
+1. I'm going to push some minor tweaks (removing some duplicated bits) and then merge once the build runs

**Pull request closed without merging**